### DISCUSSION TITLE
First thoughts on VX

### DISCUSSION THREAD

**Discussion by le-rou** [VentuzTechnology/Ventuz.Extension.prerelease#20] [General]
I wrote a first VX node and must say that I like the idea of creating nodes. I initially wanted to do this ten years ago and quickly realized that scripts wouldn't work that way. I like the idea of extending Ventuz with basic functionality, I really do! Also, the ability to create resources like geometry is great. Here are a few thoughts. Feel free to extract any point into a seperate discussion.

### VX as Scripts replacement
I use a lot of scripting for logic and as I understood it, VX is supposed to replace Scripts at some point. Currently, I can't see that. For that we'll first need to be able to reload the assemblies on the fly. When you code a script for your scene logic, you're usually working on the code and the scene at the same time, adapting the code to changes in the scene. Having to restart Ventuz breaks this workflow. But you already said that with a switch to .net 6 this would be possible. Until then, VX is useful for basic nodes, but not for scene logic, which is fine! As long as scripts remain available.

### Coding style
VX uses a lot of implicit definitions. Once you get used to coding VX, this might be handy. However, it also makes for a steeper learning curve as you must first understand all those naming conventions and what they lead to. Scripts are far more explicit in that matter.

### Deployment
The most important aspect will be the deployment. I know this is not sorted out, yet. However, it's very important that you can still export a VPR that will run on a clean Ventuz install. It must not be necessary to install additional packages. Use cases are clients who are able to copy&paste a VPR and start it - but that's it. We can't make them manage packages on their machines.

### Version management
I'm curious to see how version management will work. I guess, nuget might help here. However, we're already having enough trouble with Ventuz version management. For each project, we need to check the Ventuz version and install the correct Ventuz release. One mistake and you have to start your work all over again as a scene can't be stored in a previous version. Plugins should not add to that.

### Licensing
If/when licensing gets implemented, we need warnings during VPR export. Even better would be per-project package management. I'm not sure if the nuget package management was supposed to by system-wide or poroject-wide. However, the danger is that you get used to certain plugins that you got for yourself and then you send out a VPR and the client can't run it. It would probably a good idea to license plugins for designer, not for each runtime.

**Comment by gielians** [comment]
I think creating custom nodes might be one of the biggest features (at least for me) since the layer system was introduced. I really love the way of giving us (the developers) more possibilities to to work around or even avoid the limitations Ventuz has (and we struggle with every day). But of course it comes with a few necessities to be able to use it in production.

### Development Speed

One way we (and many others) use Ventuz mainly for fast results. Build a prototype/klick dummy, prettify it: Done. 
This is possible because Ventuz can be very fast to develop with. This sould be kept intact with the development for custom nodes. As rou described: `Having to restart Ventuz breaks this workflow.` and I am 100% with you. I don't see the problem with implicit definitions, as long as everything is well documented, which it already is. Maybe I'll find some lacks if I have more time to play around...

### Deployment

see Rou
plus: We use a different deployment method for Uniqvue where we package parts of the Ventuz Projekt folder and use director mode to run the Project. Everything the extensions need to run should at least be able to be located inside the project folder to keep the "one folder contains all data" approach Ventuz has now.

### Licensing

Not a big deal for us, yet. Maybe the per dongle "Locking"-Feature is a way to protect the code. We use it for the Uniqvue Scenes anyway.

### Version management
see Rou

### Other Stuff
I have not looked into the resource handling stuff, yet. I hope I'll find some time tomorrow. But this could be one of the biggest improvements. I would like to be able to "store" and "access" object libraries containing all kinds of data.

I often use micro scenes to handle the expandability Ventuz doesn't have. For example I build a construct of different scenes and sub-scenes to create forms with input and output all depending on a Json requested by an API. The API defines the look, input features and output to PowerPoint (yes customer is king) of a whole survey held in Ventuz. This comes with a lot of hassle and workarounds as you might imagine. Being able to cut this construct down to a few or even a single node would be awesome. This is why I would love the touch feature: Interactive expandable Charts? No Problem, just one custom Node.

**Reply by gielians** [reply]
One big point I discovered today:

### Assembly loading

Yes it is possible to load assemblies now but it is as inconvenient as it can be...
First of all they have to be put next to the Ventuz.exe inside the program folder
And I think the version control inside the main Ventuz project is set to "strict" which for dependencies like >=v1.0.0 results in Version=1.0.0 which is bad because the depended assembly maybe used elsewhere but in version >=5.0.0. Which should not be a problem if the >= would not be treated as =.
Using workarounds like [AppDomain.AssemblyResolve](https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.assemblyresolve?view=netframework-4.8) don't work because of reflection.

Example:

`Microsoft.AspNetCore.Connections.Abstractions` has a dependency of `Microsoft.Bcl.AsyncInterfaces` Version >=1.0.0.0 When running Ventuz it looks for exactly =1.0.0.0. Which is not the version loaded (5.0.0.0). 

```
17:40 26.10.2022 Error  : ServerConnector3.0: Error connecting to server: System.IO.FileLoadException: Die Datei oder Assembly "Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" oder eine Abhängigkeit davon wurde nicht gefunden. Die gefundene Manifestdefinition der Assembly stimmt nicht mit dem Assemblyverweis überein. (Ausnahme von HRESULT: 0x80131040)
Dateiname: "Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51"
   bei Microsoft.AspNetCore.SignalR.Client.HubConnection.StartAsyncCore(CancellationToken cancellationToken)
   bei Microsoft.AspNetCore.SignalR.Client.HubConnection.<StartAsyncInner>d__50.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei Microsoft.AspNetCore.SignalR.Client.HubConnection.<StartAsync>d__49.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei SessionManager.Shared.PresentationConnection.<ConnectToPresentationHubAsync>d__22.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei SessionManager.Shared.Tools.<RunTaskAsyncWithCallback>d__0.MoveNext()

=== Zustandsinformationen vor Bindung ===
LOG: DisplayName = Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
 (Fully-specified)
LOG: Appbase = file:///C:/Program Files/Ventuz Technology Group/Ventuz 6.12.0.219/
LOG: Ursprünglicher PrivatePath = NULL
Aufruf von Assembly : Microsoft.AspNetCore.Connections.Abstractions, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60.
===
LOG: Diese Bindung startet im default-Load-Kontext.
LOG: Die Anwendungskonfigurationsdatei wird verwendet: C:\Program Files\Ventuz Technology Group\Ventuz 6.12.0.219\Ventuz.exe.Config
LOG: Die Hostkonfigurationsdatei wird verwendet: 
LOG: Die Computerkonfigurationsdatei von C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config wird verwendet.
LOG: Verweis nach der Richtlinie: Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
LOG: Download von neuem URL file:///C:/Program Files/Ventuz Technology Group/Ventuz 6.12.0.219/Microsoft.Bcl.AsyncInterfaces.DLL.
WRN: Der Vergleich des Assemblynamens führte zum Konflikt: Hauptversion.
ERR: Das Setup der Assembly konnte nicht abgeschlossen werden (hr = 0x80131040). Die Suche wurde beendet.
```


I can only repeat my point stated at "Deployment": There should be a way to manage assemblies like in scripts by putting them into the project folder.

--

I played around with the Custom Resource stuff today and must say it is pretty nice and can change our standard Data-Distribution via Json strings and Json-Parsers to faster object based approaches: very nice.

**Comment by SFuhrmann** [comment]
Want to add some thoughts here while playing around and thought that extending this thread makes more sense than creating another one:

### Coding Style

I would love to have *Visual Studio Code and Visual Studio extensions* to make development with Vx a lot faster:
- *Project Templates/Project Creation Wizard* - similar to what we have with Director Plugins. Creating a template folder structure and creating proper settings for deployment, e.g. a post-build command for code validation.
- *Autocomplete* features, especially for Ctrl+Space suggestions, so that you only need to memorize which implicit keywords there are for method names etc. but do not need to remember them in detail. If possible I would also love to have an autocomplete on parameter-declaration for intermediate properties.

EDIT: Refactoring is also a problem with the naming convention approach. If you want to change the name of an intermediate property you have to manually adapt anything referencing it, because intellisense cannot do that. Is there something that could help us with that? The same goes for Attributes that reference Members - e.g. VxDefaultValueAttribute.

### Deployment

Basically we would need what we already have with Scripts: System-wide deployment (this is what we have already) and *Project-specific deployment*, the latter will care about including all the necessary files in a VPR so that you do not have to set anything up on the target machines just for your extension to work.

### Roadmap, Future Features

Most parts of the Roadmap make sense to me, though looking at it, here are some quick thoughts:
- Access to the Scene/Layer Tree, especially to create instances of objects will make a lot of use cases much easier. Not sure if its on your roadmap already, but can we also consider any kind "prefabs", e.g. through *accessing a Ventuz Node Repository*. I think of stuff like infinite/dynamic scrollers, data-driven graphics etc. These methods might need to support multi-threading or another kind of asynchronous validation.
- I don't think that *Data-Portals* are really necessary here - as far as I can tell, you could use them to make the Designer's layered UI structure more accessible to new users - who will probably not use Vx.

Other than that I really think you should push the .NET 6 migration to make Vx usable in Ventuz workflows - the biggest hurdle currently is the need to restart Ventuz each time you change anything - this makes debugging your code very painful currently.

**Comment by le-rou** [comment]
hey. It's been a while. I've actually been using VX in a few projects I did for customers. Custom geometry mostly. The issues I mentioned are really the problems I'm facing: turn around times are terrible with having to restart Ventuz, Distribution is a pain, too. It's become the number one issue: customer calls: I see the scene, but xyz is missing. I ask for the log: VX plugin not found.. forgot it again. So it would be great if the plugin could reside in the Project folder (in modules for instance).

My question is: **Is it still worth investing time here?** It seems like nothing much has happened here in the last few months and I wonder if all those great things are still on the road map? I could imaging a bit of a hen-egg problem: development rests as few people use VX, but only a few people use VX as they see it as too much of a beta. So what's the word on this?

I know a few things would be major work: like using nuget for plugin management. But again and adding to gielians: putting the plugin into the project directory would be enough. That would also help with the version management, as we maintain version control for projects anyway. So, for now, I'd have a few things that I see as very important:

- Plugin in Project Directory
- Cleanup plugin content (each plugin comes with all the dependencies, would be nice to only have the plugin DLL there)
- Workflow without restarting Ventuz

This would absolutely be enough to use small plugins in projects more often. Right now, I'm still sticking to scripts mostly, as I mentioned I use VX mostly for geometry generation (which is already an awesome benefit).

**Reply by le-rou** [reply]
unfortunately, I kind of feel like I got my answer.

**Reply by VentuzRalfStanke** [reply]
Hi Rou,

as Karol already said, we are quite busy with other high prio stuff, but let me answer some of your questions:

* putting the plugins into the project is not possible until we are not running under .net Core 7. We can't unload loaded assemblies from the current process. Runtime and Designer can switch projects on-the-fly, but we can't unload the Vx-code... so switch would become impossible
* clean packages with correctly dependencies will be solved with the integration of nuget. So we don't want to waste time on an interim solution
* see above... as long we can't unload assemblies from memory the only way is to restart the process.

We are currently converting Ventuz to run on Core 7 - this take time! the full testing is very complex since we must change a massive amount of code and external dependencies. It is not clear yet, when the Core 7 version is ready to be released. Until then, the deployment of Vx plugins will stay as it is.

We changed the API internally (not yet released with 7.00) this hopefully will be the final design. There is not much left to do and we hope we can release it soon. These changes will make you vx-code incompatible - but adjusting it is quite easy.

Sorry for this delayed reply

Ralf

**Comment by VentuzKarolStrozynski** [comment]
Hi Rou,

Finalizing the VX API is still on our road map but we currently have a lot of higher prioritized ToDos.

Best Regards
Karol
